From a588758a5ebd43790e902e8e4503b03ffb8a9f66 Mon Sep 17 00:00:00 2001
From: Jian Chen <Jian.Chen@analog.com>
Date: Fri, 24 Aug 2018 02:46:55 -0400
Subject: [PATCH] [LINUX-1427] Enable PHY DP83867 for sc589-ezkit version 2.0
 board

Starting from board version 2.0, sc589-ezkit is using new PHY DP83867, this
patch is to:
- Enable New PHY DP83867 for SC589-EZKIT version 2.0 board
- Configure the INT/PWDN pin on DP83867 as an Interrupt Output instead of a
Power-Down Input on ADSP-SC589-EZKIT version 2.0 board in order to prevent the
signal interference from other peripherals during they are running at the same
time.

Signed-off-by: jian chen <jian.chen@analog.com>
---
 board/adi/sc589-ezkit/sc589-ezkit.c | 28 +++++++++++++++++++++++-----
 include/configs/sc589-ezkit.h       |  2 +-
 2 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/board/adi/sc589-ezkit/sc589-ezkit.c b/board/adi/sc589-ezkit/sc589-ezkit.c
index 5999271..09c6897 100644
--- a/board/adi/sc589-ezkit/sc589-ezkit.c
+++ b/board/adi/sc589-ezkit/sc589-ezkit.c
@@ -10,6 +10,7 @@
 #include <netdev.h>
 #include <phy.h>
 #include <asm/io.h>
+#include <asm/gpio.h>
 #include <asm/mach-types.h>
 #include <asm/arch/portmux.h>
 #include <asm/arch/sc58x.h>
@@ -82,9 +83,15 @@ int board_eth_init(bd_t *bis)
 
 	if (CONFIG_DW_PORTS & 1) {
 		gpio_request(GPIO_PB14, "emac0_phy_reset");
+		gpio_request(GPIO_PC15, "emac0_phy_pwdn");
+		gpio_direction_output(GPIO_PC15, 1);
 		gpio_direction_output(GPIO_PB14, 1);
+		mdelay(20);
+		gpio_direction_output(GPIO_PB14, 0);
+		mdelay(90);
+		gpio_direction_output(GPIO_PB14, 1);
+		mdelay(20);
 
-		mdelay(1);
 		writel((readl(REG_PADS0_PCFG0) | 0xc), REG_PADS0_PCFG0);
 
 		static const unsigned short pins[] = P_RGMII0;
@@ -107,10 +114,21 @@ int board_phy_config(struct phy_device *phydev)
 	int  phy_data = 0;
 
 	if (CONFIG_DW_PORTS & 1) {
-		phy_data = phy_read(phydev, MDIO_DEVAD_NONE, 0x12);
-
-		/* enable 3com mode for RGMII */
-		phy_write(phydev, MDIO_DEVAD_NONE, 0x12, (3 << 12) | phy_data);
+		phy_data = phy_read(phydev, MDIO_DEVAD_NONE, 0x32);
+		phy_write(phydev, MDIO_DEVAD_NONE, 0x32, (1 << 7) | phy_data);
+#ifdef CONFIG_PHY_TI
+		int cfg3 = 0;
+		#define MII_DP83867_CFG3    (0x1e)
+		/*
+		 * Pin INT/PWDN on DP83867 should be configured as an Interrupt Output
+		 * instead of a Power-Down Input on ADI SC5XX boards in order to
+		 * prevent the signal interference from other peripherals during they
+		 * are running at the same time.
+		 */
+		cfg3 = phy_read(phydev, MDIO_DEVAD_NONE, MII_DP83867_CFG3);
+		cfg3 |= (1 << 7);
+		phy_write(phydev, MDIO_DEVAD_NONE, MII_DP83867_CFG3, cfg3);
+#endif
 	}
 
 	if (CONFIG_DW_PORTS & 2)
diff --git a/include/configs/sc589-ezkit.h b/include/configs/sc589-ezkit.h
index adde83a..0988644 100644
--- a/include/configs/sc589-ezkit.h
+++ b/include/configs/sc589-ezkit.h
@@ -118,13 +118,13 @@
 #define CONFIG_DTBNAME		"sc589-ezkit.dtb"
 #define CONFIG_HOSTNAME		"sc58x"
 #define CONFIG_DESIGNWARE_ETH
-#define CONFIG_PHY_ADDR		1
 #define CONFIG_DW_PORTS		1
 #define CONFIG_DW_AUTONEG
 #define CONFIG_DW_ALTDESCRIPTOR
 #define CONFIG_DW_AXI_BURST_LEN 16
 #define CONFIG_MII
 #define CONFIG_PHYLIB
+#define CONFIG_PHY_TI
 #define CONFIG_ETHADDR	02:80:ad:20:31:e8
 
 /*
-- 
1.9.1

